text/0000-pack-package-refactor.md

```diff
@@ -0,0 +1,58 @@
+# Meta
+[meta]: #meta
+- Name: (fill in the feature name: My Feature)
+- Start Date: 02/24/21
+- Author(s): dwillist
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a restructuring of the `pack` codebase into reusable modular packages.
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+    - Maintaining a modular and re-usable codebase enables contributors to be more effective, and encourages new adopters. 
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC is looking accomplish the following:
+- enable pack library consumers by moving code out of `internal`
+- ease the process of contribution by having well defined connectable components.
+- Follow a general process of if data is externally visible on a pack produced artifact. We should provide structs for decoding.
+
+To accomplish the above we propose the following package structure:
+
+```
+.
+├── cmd/
+│   ├── pack/ → current pack cli
+│   └── fetcher/ → cli to download buildpacks
```

Comment:
This seems like a brand new feature rather than a refactor. Maybe this should be out of scope.

Reply:
It is my understanding that this structures is based on some [future thinking hypotheticals](https://github.com/buildpacks/community/discussions/51#discussion-1608992). I'd propose keeping it in the RFC but making note that it's "future thinking".

```suggestion
│   └── fetcher/ → cli to download buildpacks (future thinking)
```

Reply:
Personally I would rather cut it from this RFC. I have misgivings about breaking `pack` into smaller binaries and believe worthy of its own RFC.

The `(future thinking)` caveat makes me feel slightly better, but also I don't see a reason to include this at all. The question of where we would put future CLIs doesn't seem like a controversy that need resolution (they always go in `cmd` by convention).

Reply:
I think re-framing the goals of this rfc as focusing on a process to pull functionality out means we can defer this conversation until we need to distribute additional CLIs